File: advent-of-code/2015/01/01/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

func calculateFloor(source string) int {
	floor := 0
	for _, char := range source {
		switch char {
		case '(':
			floor++
		case ')':
			floor--
		}
	}
	return floor
}

func main() {
	// Test cases with assertions
	if calculateFloor("(())") != 0 {
		panic("Test failed: (())")
	}
	if calculateFloor("()()") != 0 {
		panic("Test failed: ()()")
	}
	if calculateFloor("(((") != 3 {
		panic("Test failed: (((")
	}
	if calculateFloor("(()(()(") != 3 {
		panic("Test failed: (()(()(")
	}
	if calculateFloor("))(((((") != 3 {
		panic("Test failed: ))(((((")
	}
	if calculateFloor("())") != -1 {
		panic("Test failed: ())")
	}
	if calculateFloor("))(") != -1 {
		panic("Test failed: ))(")
	}
	if calculateFloor(")))") != -3 {
		panic("Test failed: )))")
	}
	if calculateFloor(")())())") != -3 {
		panic("Test failed: )())())")
	}

	// Calculate and print the answer
	_, filename, _, _ := runtime.Caller(0)
	scriptDir := filepath.Dir(filename)
	inputPath := filepath.Join(scriptDir, "..", "input.txt")
	input, _ := os.ReadFile(inputPath)
	fmt.Printf("Go: %d\n", calculateFloor(string(input)))
}

File: advent-of-code/2015/01/01/solution.rs
use std::fs;

fn calculate_floor(source: &str) -> i32 {
    source.chars().fold(0, |floor, character| match character {
        '(' => floor + 1,
        ')' => floor - 1,
        _ => floor,
    })
}

fn main() {
    // Test cases with assertions
    assert_eq!(calculate_floor("(())"), 0);
    assert_eq!(calculate_floor("()()"), 0);
    assert_eq!(calculate_floor("((("), 3);
    assert_eq!(calculate_floor("(()(()("), 3);
    assert_eq!(calculate_floor("))((((("), 3);
    assert_eq!(calculate_floor("())"), -1);
    assert_eq!(calculate_floor("))("), -1);
    assert_eq!(calculate_floor(")))"), -3);
    assert_eq!(calculate_floor(")())())"), -3);

    // Calculate and print the answer
    let script_dir = std::path::Path::new(file!()).parent().unwrap();
    let input_path = script_dir.join("..").join("input.txt");
    let input = fs::read_to_string(input_path).unwrap();
    println!("Rust: {}", calculate_floor(&input));
}
